neuralmonkey/runners/beamsearch_runner.py

```diff
@@ -1,68 +1,160 @@
+            self.parent_ids,
+            bs_outputs.last_search_step_output.parent_ids[0:step],
+            axis=0)
+
+        self.token_ids = np.append(
+            self.token_ids,
+            bs_outputs.last_search_step_output.token_ids[0:step],
+            axis=0)
+
+        if (self._decoder.max_output_len is not None and
+                self.step > self._decoder.max_output_len):
+            self.prepare_results()
+            return
+
+        # Prepare the next feed_dict (in ensembles)
+        self._next_feed = []
+        for result in results:
+            bs_outputs = result["bs_outputs"]
+
+            search_state = bs_outputs.last_search_state._replace(
+                prev_logprobs=ens_logprobs)
+
+            fd = {self._decoder.max_steps: 1,
+                  self._decoder.search_state: search_state}
+
+            dec_feedables = bs_outputs.last_dec_loop_state
+
+            # Due to the arrays in DecoderState (prev_contexts),
+            # we have to create feed for each value separately.
+            for field in self._decoder.decoder_state._fields:
+                # We do not feed the step
+                if field == "step":
+                    continue
+                tensor = getattr(self._decoder.decoder_state, field)
+                value = getattr(dec_feedables, field)
+                if isinstance(tensor, list):
+                    for t, val in zip(tensor, value):
+                        fd.update({t: val})
+                else:
+                    fd.update({tensor: value})
+
+            self._next_feed.append(fd)
+
+        if self.token_ids.size == 0:
+            return
+
+        # We assume that we can stop decoding when all tokens
+        # in the last step were <pad>
+        if np.all(np.equal(self.token_ids[-1], PAD_TOKEN_INDEX)):
+            self.prepare_results()
+    # pylint: enable=too-many-locals
+
+    def prepare_results(self):
+        max_time = self.step
+
+        output_tokens = []
+        hyp_idx = np.argpartition(
+            -self.scores[-1], self._rank - 1)[self._rank - 1]
+        bs_score = self.scores[-1][hyp_idx]
         for time in reversed(range(max_time)):
-            token_id = evaluated_bs.token_ids[time][hyp_index]
-            token = self._vocabulary.index_to_word[token_id]
+            token_id = self.token_ids[time][hyp_idx]
+            token = self._decoder.vocabulary.index_to_word[token_id]
             output_tokens.append(token)
-            hyp_index = evaluated_bs.parent_ids[time][hyp_index]
+            hyp_idx = self.parent_ids[time][hyp_idx]
 
         output_tokens.reverse()
 
-        before_eos_tokens = []  # type: List[str]
+        before_eos_tokens = []
         for tok in output_tokens:
             if tok == END_TOKEN:
                 break
-            before_eos_tokens.append(tok)
+            # TODO: investigate why the decoder can start generating
+            # padding before generating the END_TOKEN
```

Comment:
podle mě proto, že tenhle kód obchází ty runtime logprobs toho dekoderu kde se to vylouci automaticky, ale kdyz se ensemble shodne, ze chce pad, tak s tim nic nenadelas. stejnej trik by se asi dal udelat nekde kolem radku 60-62.

nebo to je jinak?

Reply:
Nemyslim si, ze by kterykoliv z modelu mel do ensemblu tlacit pad, pokud na predchozim vstupu nedostal pad nebo eos. Predchozi vstup je ale stejny pro vsechny modely z ensemblu.

Ale je mozne, ze mi neco unika.

Reply:
ty modely nejsou trénovaný na to, aby nevyhazovaly pad. ty padding logity se nesmí používat, protože jsou víceméně náhodný. takže se občas může stát, že je pad logit největší. Ta část dekodérů, která ten padding zakazuje, se tady nepoužívá